modules/PatternUtils.js

```diff
@@ -1,79 +1,26 @@
-import invariant from 'invariant'
+import pathToRegexp from 'path-to-regexp'
 
-function escapeRegExp(string) {
-  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
-}
-
-function escapeSource(string) {
-  return escapeRegExp(string).replace(/\/+/g, '/+')
-}
-
-function _compilePattern(pattern) {
-  let regexpSource = ''
-  const paramNames = []
-  const tokens = []
-
-  let match, lastIndex = 0, matcher = /:([a-zA-Z_$][a-zA-Z0-9_$]*)|\*\*|\*|\(|\)/g
-  while ((match = matcher.exec(pattern))) {
-    if (match.index !== lastIndex) {
-      tokens.push(pattern.slice(lastIndex, match.index))
-      regexpSource += escapeSource(pattern.slice(lastIndex, match.index))
-    }
-
-    if (match[1]) {
-      regexpSource += '([^/?#]+)'
-      paramNames.push(match[1])
-    } else if (match[0] === '**') {
-      regexpSource += '([\\s\\S]*)'
-      paramNames.push('splat')
-    } else if (match[0] === '*') {
-      regexpSource += '([\\s\\S]*?)'
-      paramNames.push('splat')
-    } else if (match[0] === '(') {
-      regexpSource += '(?:'
-    } else if (match[0] === ')') {
-      regexpSource += ')?'
-    }
+const RegexpCache = {}
+const CompiledCache = {}
 
-    tokens.push(match[0])
-
-    lastIndex = matcher.lastIndex
-  }
-
-  if (lastIndex !== pattern.length) {
-    tokens.push(pattern.slice(lastIndex, pattern.length))
-    regexpSource += escapeSource(pattern.slice(lastIndex, pattern.length))
+function getRegexp(pattern) {
+  if (!RegexpCache[pattern]) {
+    RegexpCache[pattern] = pathToRegexp(pattern, { end: false })
   }
 
-  return {
-    pattern,
-    regexpSource,
-    paramNames,
-    tokens
-  }
+  return RegexpCache[pattern]
 }
 
-const CompiledPatternsCache = {}
-
-export function compilePattern(pattern) {
-  if (!(pattern in CompiledPatternsCache))
-    CompiledPatternsCache[pattern] = _compilePattern(pattern)
+function getCompiled(pattern) {
+  if (!CompiledCache[pattern]) {
+    CompiledCache[pattern] = pathToRegexp.compile(pattern)
+  }
 
-  return CompiledPatternsCache[pattern]
+  return CompiledCache[pattern]
 }
 
 /**
- * Attempts to match a pattern on the given pathname. Patterns may use
- * the following special characters:
- *
- * - :paramName     Matches a URL segment up to the next /, ?, or #. The
- *                  captured string is considered a "param"
- * - ()             Wraps a segment of the URL that is optional
- * - *              Consumes (non-greedy) all characters up to the next
- *                  character in the pattern, or to the end of the URL if
- *                  there is none
- * - **             Consumes (greedy) all characters up to the next character
- *                  in the pattern, or to the end of the URL if there is none
```

Comment:
I might leave this comment block in place, updated to the new syntax. Either that or link to the path-to-regexp docs.